src/exe/Main.hs

```diff
@@ -225,17 +248,41 @@ processInputFile
   :: Members '[Report, Embed IO] r => Options -> FilePath -> Sem r ()
 processInputFile opts inputFile = do
   input <- embed $ readFile inputFile
-  let state       = initPMState opts
-      inputModule = HSE.fromParseResult (HSE.parseModule input)
-      intermediateModule =
-        evalPM (processModule (FromHSE.transformModule inputModule)) state
-      outputModule = ToHSE.transformModule inputModule intermediateModule
+  let state             = initPMState opts
+      (output, modName) = processInput input state (optFrontend opts)
+--      inputModule = HSE.fromParseResult (HSE.parseModule input
+--      intermediateModule =
+--        evalPM (processModule (FromHSE.transformModule inputModule)) state
+--      outputModule = ToHSE.transformModule inputModule intermediateModule
   case optOutputDir opts of
     Just outputDir -> do
-      let outputFile = outputDir </> makeOutputFileName inputFile inputModule
+      let outputFile = outputDir </> makeOutputFileName inputFile modName
       embed $ createDirectoryIfMissing True (takeDirectory outputFile)
-      embed $ writeFile outputFile (prettyPrintModule outputModule)
-    Nothing -> embed $ putStrLn (prettyPrintModule outputModule)
+      embed $ writeFile outputFile output -- (prettyPrintModule outputModule)
+    Nothing -> embed $ putStrLn output --(prettyPrintModule outputModule)
+
+-- | Processes a given module as a string with the corresponding frontend.
+--   Returns the output as a string and the module name.
+processInput
+  :: String -> PMState FromHSE.HSE -> Frontend -> (String, Maybe String)
```

Comment:
This is not actually parameterized over the front end with this signature. Check if the following works:
```suggestion
processInput
  :: String -> PMState a -> Frontend -> (String, Maybe String)
```
If not, you have to restructure a bit.